New blog post on Swagger UI (#803)
quarkusio/quarkusio.github.io: _posts/2020-11-25-stylish-api.adoc

=== Theme

https://ostranme.github.io/swagger-ui-themes/[Swagger UI Themes] are now available in configuration, with the default theme being 'feeling blue'.

You can change the theme by setting the `quarkus.swagger-ui.theme` property, for example:

[source,properties]
----
quarkus.swagger-ui.theme=outline
----

image::themed.png[themed]

You can also go back to the original (vanilla) Swagger UI theme:

[source,properties]
----
quarkus.swagger-ui.theme=original
----

image::original.png[original]

Theme options available:

* feeling-blue (default)
* original
* flattop
* material
* monokai
* muted
* newspaper
* outline

=== Logo

As part of the custom branding, you can supply your own logo to replace the Quarkus logo. For example, let say you own company that makes everything, ACME, and you are using REST Services for your online store, and wants to brand the Swagger UI Page:

image::acme_logo.png[acme logo]

NOTE: Hot reload is not working for logo changes, and remember browser cache, you might need to https://refreshyourcache.com/en/cache/[force refresh] your browser.

> To supply your own logo, you need to place a file called `logo.png` in `src/main/resources/META-INF/branding`.

Author: I have no strong feeling on this. We can change this.

> To supply your own logo, you need to place a file called `logo.png` in `src/main/resources/META-INF/branding`.

Reviewer: What I don't like is that there's a good chance the logo version used here might be a bit specific to these UIs and `branding/logo.png` is typically where I would put a very generic logo.

Now we would need a name to cover all these UIs.

> To supply your own logo, you need to place a file called `logo.png` in `src/main/resources/META-INF/branding`.

Author: This logo will be used for all UIs.
What we can do later is allow a more specific logo per UI, calling it something like "swagger-ui.png" or "graphql-ui.png" or something like that. 
The style.css already works like that. (I did not mention that in the blog post)

> To supply your own logo, you need to place a file called `logo.png` in `src/main/resources/META-INF/branding`.

Author: So with style, it will use the specific one (per UI) or fallback to style.css. We can do the same for logo.

=== Style

You can go further, and supply your own `style.css`, to fine-tune the branding. In example, to change the `topbar` of the Swagger-UI screen to the corporate colors of ACME:

[source,css]
----
html{
    box-sizing: border-box;
    overflow: -moz-scrollbars-vertical;
    overflow-y: scroll;
}

*,
*:before,
*:after
{
    box-sizing: inherit;
}

body{
    margin:0;
    background: white;
}

.swagger-ui .topbar {
    background-color: whitesmoke; <1>
}

#footer {
    background-color: whitesmoke;
    font-family:sans-serif;
    color:#4da32c;
    font-size:70%;
    text-align: center;
}
----

<1> here set the `topbar` background color.

image::acme_css.png[acme css]

> You can change any styling element in this css file, you need to place this file called `style.css` in `src/main/resources/META-INF/branding`.

Reviewer: Same here about the name being very generic.

=== Other styling options

You can also set the HTML title, and add a footer:

[source,properties]
----
quarkus.swagger-ui.title=ACME API
quarkus.swagger-ui.footer=&#169; 2020 . ACME
----

> Along with other OpenAPI Header fields that can be set via config (as discussed in link:/blog/openapi-for-everyone/[this post]):

Reviewer: ```suggestion
Along with other OpenAPI Header fields that can be set via configuration properties (as discussed in link:/blog/openapi-for-everyone/[this post]):
```

[source,properties]
----
mp.openapi.extensions.smallrye.info.title=ACME online store API
mp.openapi.extensions.smallrye.info.version=1.0.0
mp.openapi.extensions.smallrye.info.description=We make everything, and sell it online
mp.openapi.extensions.smallrye.info.contact.email=it@acme.com
mp.openapi.extensions.smallrye.info.contact.name=ACME IT
mp.openapi.extensions.smallrye.info.contact.url=https://www.acme.com
mp.openapi.extensions.smallrye.info.license.name=Apache 2.0
mp.openapi.extensions.smallrye.info.license.url=http://www.apache.org/licenses/LICENSE-2.0.html
----

The UI is now fully branded:

image::acme_footer.png[acme footer]

== Other Swagger UI Options

> Another new feature available in Quarkus (v1.10.0 +) is the ability to set any of the https://swagger.io/docs/open-source-tools/swagger-ui/usage/configuration/[configuration options] available in Swagger UI. As an example, we can set the `urls` and add the petstore (as the default selected option) to Swagger UI:

Reviewer: ```suggestion
Another new feature available in Quarkus (version 1.10+) is the ability to set any of the https://swagger.io/docs/open-source-tools/swagger-ui/usage/configuration/[configuration options] available in Swagger UI. As an example, we can set the `urls` and add the petstore (as the default selected option) to Swagger UI:
```

[source,properties]
----
quarkus.swagger-ui.urls.default=/openapi
quarkus.swagger-ui.urls.petstore=https://petstore.swagger.io/v2/swagger.json
quarkus.swagger-ui.urls-primary-name=petstore
----

This will change the `topbar` to have a dropdown box with the urls provided:

image::petstore.png[petstore]

Another example, `supportedSubmitMethods` can hide the `Try it out` button for certain HTTP Method Types:

[source,properties]
----
quarkus.swagger-ui.supported-submit-methods=get
----

Note below the missing `Try it out` button on the `POST`

image::tryitout.png[try it out]

> All other Swagger UI options are now available to configure the UI.

Reviewer: ```suggestion
All the other Swagger UI options are now available to configure the UI.
```

== Other small new features

> Two small new features in OpenAPI and Swagger UI, the ability to add the Health Endpoints and the ability to disable the UI and/or Schema in Runtime.

Reviewer: ```suggestion
Let me present you two small new features in OpenAPI and Swagger UI: the ability to add the Health Endpoints and the ability to disable the UI and/or Schema at runtime.
```

=== Add Health API to Open API

If you are using the `smallrye-health` extension, you can add the Health Endpoints to OpenAPI:

[source,properties]
----
quarkus.health.openapi.included=true
----

image::health.png[health]

> === Disable in Runtime

Reviewer: ```suggestion
=== Disable at Runtime
```

If you included the UI in your app (`quarkus.swagger-ui.always-include=true`), you can now disable it when starting the application.

> ----

Reviewer: ```suggestion
[source]
----
```
